scvi/module/_lda.py

```diff
@@ -0,0 +1,299 @@
+import math
+from typing import Dict, Iterable, Optional, Union
+
+import numpy as np
+import pyro
+import pyro.distributions as dist
+import torch
+import torch.nn as nn
+import torch.nn.functional as F
+from pyro.infer import Trace_ELBO
+from pyro.nn import PyroModule
+
+from scvi._constants import _CONSTANTS
+from scvi.module.base import PyroBaseModuleClass, auto_move_data
+from scvi.nn import FCLayers
+
+_LDA_PYRO_MODULE_NAME = "lda"
+
+
+class CategoricalBoW(dist.Multinomial):
+    def log_prob(self, value):
+        if self._validate_args:
+            self._validate_sample(value)
+        logits, value = dist.util.broadcast_all(self.logits, value)
+        logits = logits.clone(memory_format=torch.contiguous_format)
+        logits[(value == 0) & (logits == -math.inf)] = 0
+        log_powers = (logits * value).sum(-1)
+        return log_powers
+
+
+class LDAPyroModel(PyroModule):
+    """
+    A PyroModule that serves as the model for the LDAPyroModule class.
+
+    Parameters
+    ----------
+    n_input
+        Number of input genes.
+    n_components
+        Number of components/topics to model.
+    cell_component_prior
+        Prior of cell component distribution.
+    component_gene_prior
+        Prior of component gene distribution.
+    """
+
+    def __init__(
+        self,
+        n_input: int,
+        n_components: int,
+        cell_component_prior: float,
+        component_gene_prior: float,
+    ):
+        super().__init__(_LDA_PYRO_MODULE_NAME)
+
+        self.n_input = n_input
+        self.n_components = n_components
+        # Populated by PyroTrainingPlan.
+        self.n_obs = None
+
+        self.register_buffer(
+            "cell_component_prior",
+            torch.full((self.n_components,), cell_component_prior),
+        )
+        self.register_buffer(
+            "component_gene_prior", torch.full((self.n_input,), component_gene_prior)
+        )
+
+        # Hack: to allow auto_move_data to infer device.
+        self._dummy = torch.nn.Parameter(torch.zeros(1), requires_grad=False)
```

Comment:
it's a little cleaner to register a buffer here. This only works because our optimizer filters params based on the requires grad flag

Reply:
but automove only iterates through model parameters. register_buffer doesn't add it to the parameters. Otherwise, this line wouldnt' be necessary since the previous lines already registered some stuff